File: src/leetcode/python/easy/HappyNumber_202.py
class HappyNumber_202(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        def calculate(num):
            return sum([int(i) ** 2 for i in str(num)])

        slow = fast = n
        while True:
            slow = calculate(slow)
            fast = calculate(calculate(fast))
            if fast == 1:
                return True
            if fast == slow:
                return False

File: src/leetcode/java/HappyNumber_202.java
package leetcode.java;

public class HappyNumber_202 {
  public boolean isHappy(int n) {
    int slow = n, fast = n;
    do {
      slow = calculate(slow);
      fast = calculate(calculate(fast));
      if (fast == 1) return true;
    } while (slow != fast);
    return false;
  }

  private int calculate(int n) {
    int res = 0, tmp;
    while (n > 0) {
      tmp = n % 10;
      res += tmp * tmp;
      n /= 10;
    }
    return res;
  }
}
